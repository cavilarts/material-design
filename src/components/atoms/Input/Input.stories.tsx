import { Meta, Story } from '@storybook/react/types-6-0';
import Input, { inputProps } from './index';

const inputSettings = {
  component: Input,
  title: 'Atoms/Input',
  argTypes: {
    type: {
      control: 'select',
      options: [
        'color',
        'date',
        'datetime-local',
        'email',
        'month',
        'number',
        'password',
        'search',
        'tel',
        'text',
        'time',
        'url',
        'week',
      ],
    },
    label: {
      control: 'text',
    },
    placeholder: {
      control: 'text',
    },
  },
};

export default inputSettings as Meta;

export const input: Story<inputProps> = ({ type, label, placeholder }) => {
  return <Input type={type} label={label} placeholder={placeholder} />;
};
