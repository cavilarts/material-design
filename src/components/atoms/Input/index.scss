.input {
  &::placeholder {
    color: transparent;
  }

  &--active {
    &::placeholder {
      color: gray;
    }
  }

  &__wrapper {
    position: relative;
  }

  &__labelWrapper {
    position: absolute;
    left: 0;
    width: 100%;
    height: 100%;
    box-sizing: content-box;
    pointer-events: none;
  }

  &__label {
    display: block;
    left: 0;
    position: absolute;
    transform: perspective(100px);
    transition: transform 0.2s ease;

    &--active {
      display: block;
      transform: translateY(-1em) scale(0.75) perspective(100px) translateZ(0.001px);
    }
  }
}
