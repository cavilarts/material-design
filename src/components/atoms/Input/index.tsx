import React, { SyntheticEvent, useCallback, useState } from 'react';
import classNames from 'classnames';

import './index.scss';

export interface inputProps {
  type:
    | 'color'
    | 'date'
    | 'datetime-local'
    | 'email'
    | 'month'
    | 'number'
    | 'password'
    | 'search'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week';
  label?: string;
  placeholder?: string;
  onChange?: (event: SyntheticEvent) => void;
  onKeyDown?: (event: SyntheticEvent) => void;
}

const Input: React.FC<inputProps> = ({ type, label, placeholder, onChange, onKeyDown }) => {
  const [inputFocus, setInputFocus] = useState(false);

  const onChangeCallback = useCallback(
    event => {
      if (onChange) {
        onChange(event.target.value);
      }
    },
    [onChange],
  );
  const onKeyDownCallback = useCallback(
    event => {
      if (onKeyDown) {
        onKeyDown(event);
      }
    },
    [onKeyDown],
  );

  return (
    <div className='input__wrapper'>
      <span className='input__labelWrapper'>
        {label ? (
          <label className={classNames('input__label', { 'input__label--active': inputFocus })}>{label}</label>
        ) : null}
      </span>
      <input
        className={classNames('input', { 'input--active': inputFocus })}
        type={type}
        placeholder={placeholder}
        onChange={onChangeCallback}
        onKeyDown={onKeyDownCallback}
        onFocus={() => setInputFocus(true)}
        onBlur={() => setInputFocus(false)}
      />
    </div>
  );
};

export default Input;
