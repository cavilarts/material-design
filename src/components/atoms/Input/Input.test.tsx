import { shallow, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import Input from './index';

configure({ adapter: new Adapter() });

describe('Input', () => {
  it('Should render a default input', () => {
    const input = shallow(<Input type='text' />);

    expect(input).toMatchSnapshot();
  });

  it('Should receive type property', () => {
    const input = shallow(<Input type='text' />);

    expect(input.find('input').prop('type')).toBe('text');
  });

  it('Should contain a placeholder', () => {
    const input = shallow(<Input type='text' placeholder='A placeholder value' />);

    expect(input.find('input').prop('placeholder')).toBe('A placeholder value');
  });

  it('Should handle onChange', () => {
    const onChangeCallback = jest.fn();
    const input = shallow(<Input type='text' onChange={onChangeCallback} />);

    input.find('input').simulate('change', { target: { value: 'Hello' } });
    expect(onChangeCallback).toHaveBeenCalledWith('Hello');
  });

  it('should execute onChange even if not provided', () => {
    const input = shallow(<Input type='text' />);

    input.find('input').simulate('change', { target: { value: 'Hello' } });
    expect(input.find('input').prop('value')).toBe('Hello');
  });

  it('Should handle onKeyDown', () => {
    const onKeyDownCallback = jest.fn();
    const input = shallow(<Input type='text' onKeyDown={onKeyDownCallback} />);

    input.find('input').simulate('keydown', { key: 'z', keyCode: 90, which: 90 });
    expect(onKeyDownCallback).toHaveBeenCalledWith({ key: 'z', keyCode: 90, which: 90 });
  });

  it('Should execute onKeyDown even if not provided', () => {
    const input = shallow(<Input type='text' />);

    input.find('input').simulate('keydown', { target: { value: 'z' }, key: 'z', keyCode: 90, which: 90 });
    expect(input.find('input').prop('value')).toBe('');
  });

  it('Should contain a label', () => {
    const input = shallow(<Input type='email' label='Email' />);

    expect(input.find('label').text()).toBe('Email');
    expect(input.find('label').length).toBe(1);
  });

  it('Should contain an id prop', () => {
    const input = shallow(<Input type='text' label='Email' id='email' />);

    expect(input.find('input').prop('id')).toBe('email');
    expect(input.find('label').prop('htmlFor')).toBe('email');
  });

  it('Should active the input on focus', () => {
    const input = shallow(<Input type='text' label='Email' id='email' />);

    input.find('input').simulate('focus');
    expect(input.find('label').prop('className')).toBe('input__label input__label--active');
    expect(input.find('input').prop('className')).toBe('input input--active');

    input.find('input').simulate('blur');
    expect(input.find('label').prop('className')).toBe('input__label');
    expect(input.find('input').prop('className')).toBe('input');
  });
});
