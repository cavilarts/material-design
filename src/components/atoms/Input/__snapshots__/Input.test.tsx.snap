// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Should render a default input 1`] = `ShallowWrapper {}`;
